Services/APIVersioning.md

```diff
@@ -0,0 +1,144 @@
+
+If a resource accepts particular HTTP methods, these methods should continue to be supported, and perform the same actions for as long as clients are expected to access the resource.
+
+### Do not create new required parameters or post/put data for existing resources
+
+Adding new mandatory query parameters to a request, or adding new mandatory data requirements to a PUT or POSTed representation, will break existing clients that are not aware of the mandatory information. When adding new data or query parameters to a request, API developers should assume sensible defaults for clients that are not aware of the required information.
+
+### Do not change the names or types of elements within structured data
+
+When returning structured data to the client, the type of any element within that data must not change. For example, replacing a String with an Integer, or replacing a single element with a list.
+
+### Do not change the meaning of elements within structured data
+
+Consider the example:
+
+```json
+    {
+        "name": "George",
+        "color": "red"
+    }
+```
+
+Adding a new color possibility of "blue" would not be changing the meaning of the color element. 
+
+Reclassifying resources that used to be "red" as "reddish-brown" could be changing its meaning, as clients would be affected by existing resources being reclassified, and should only be done if the aim of the change is to stop those clients mis-classifying the resource as being red when it isn't.
+
+Changing the value of color to "#FF0000" would be changing the meaning of the color element, as clients that were previously looking for "red" would start misrepresenting resources that were red as not being red.
+
+Removing the color element entirely, even for resources that are red, would be changing its meaning. This is true even if the color element was optional in the response, as clients would assume such an entity was colorless.
+
+# A Note on HATEOAS
+
+HATEOAS is the unwieldy acronym for "Hypertext As The Engine Of Application State": a standard for writing dynamically discoverable REST API where every resource exposed by the API must be reachable by retrieving the API's root resource, then following a chain of hypertext links from there. It is _highly encouraged_ that REST APIs follow HATEOAS principles, because doing so allows for more resilient APIs and more manageable API migration, and frees clients from having to hard-code the URL structure of the service.
+
+Internal clients of HATEOAS APIs must not assume any knowledge of the service's API structure that is not derived from the API itself. If a service moves resources to new locations, it should be able to expect that clients will adapt once their cached resources expire.
+
+If the API is publicly available, services must not expect that third-party clients will follow HATEOAS principles.
```

Comment:
You would still have to leave redirects at old locations to satisfy backwards compatibility, but you could expect that well-written clients would only be using those redirects for the time it took for the thing linking to the old location to expire from cache.

Similarly, a mobile client could be optimized to rely on "stale" cached resources on launch so initial response time is not affected, then perform if-modified checks on those resources in the background after it has given the user what they wanted. 

The net effect is that the default case is no slower than hard-coding the locations of those resources, while the "things have moved" case is a bit slower for the first request after things have moved, but then is self-healing.